# Set Up Paths and Filenames

```{r}
library(tidyverse)
```

```{r}
workspace_dir = "workspace"
Sys.setenv(WORKSPACE_DIR=workspace_dir)
Sys.setenv(QIIME_MAP_FILE=file.path(workspace_dir, "rat_lung_qiime_map.tsv"))
Sys.setenv(MAP_FILE=file.path("notes_and_info", "rat_lung_map.tsv"))
Sys.setenv(RAW_DATA_DIR=file.path("raw_data", "160614_McKenney_fastqs"))
Sys.setenv(TAGGED_FASTQ_DIR=file.path(workspace_dir, "tagged_fastq"))
Sys.setenv(CONDA_BIN="/opt/conda/bin")

split_fastq_base = file.path(workspace_dir, "split_fastq")
Sys.setenv(SPLIT_FASTQ_BASE=split_fastq_base)
Sys.setenv(SPLIT_FASTQ_MD5SUM=file.path(split_fastq_base, "split_fastq_md5sum.txt"))

combined_read_dir=file.path(split_fastq_base, "combined_reads")
combined_read_md5sum=file.path(combined_read_dir, "combined_fastq_md5sum.txt")

Sys.setenv(COMBINED_READ_DIR=combined_read_dir)
Sys.setenv(COMBINED_READ_MD5SUM=combined_read_md5sum)
experiment_table_file = file.path(combined_read_dir, "experiment_paired_fastq_spreadsheet.tsv")
```

SPLIT_FASTQ_BASE := $(WORKSPACE_DIR)/split_fastq
SPLIT_FASTQ_MD5SUM := $(SPLIT_FASTQ_BASE)/split_fastq_md5sum.txt
RESULTS_DIR := results
#--------------------------------------------------
#--------------------------------------------------


# READ1_FASTQ="$(RAW_DATA_DIR)/Undetermined_S0_L001_R1_001.fastq.gz"
# READ2_FASTQ="$RAW_DATA_DIR/Undetermined_S0_L001_R2_001.fastq.gz"
# INDEX_FASTQ="$RAW_DATA_DIR/Undetermined_S0_L001_I1_001.fastq.gz"

FASTQ_PREFIX := $(RAW_DATA_DIR)/Undetermined_S0_L001
FASTQ_SUFFIX := 001.fastq.gz
READ1_FASTQ := $(FASTQ_PREFIX)_R1_$(FASTQ_SUFFIX)
READ2_FASTQ := $(FASTQ_PREFIX)_R2_$(FASTQ_SUFFIX)

#--------------------------------------------------

RAW_FASTQS := $(READ1_FASTQ) $(READ2_FASTQ)
TAGGED_FASTQS := $(addprefix $(TAGGED_FASTQ_DIR)/, $(notdir $(RAW_FASTQS:.fastq.gz=_tagged.fastq)))
SPLIT_FASTQS := $(addprefix $(WORKSPACE_DIR)/split_, $(addsuffix /COMPLETION_STAMP, $(notdir $(RAW_FASTQS:.fastq.gz=))))

#--------------------------------------------------
FULL_PHYLOSEQ_RDS := $(RESULTS_DIR)/rat_lung_ps.rds
#--------------------------------------------------


all : $(TAGGED_FASTQS) $(SPLIT_FASTQS) $(FULL_PHYLOSEQ_RDS)

# Generate a MAP file that QIIME will accept

```{bash}
sed  '1s/^/#/' $MAP_FILE > $QIIME_MAP_FILE
```

# Generate tagged fastqs
```{bash}
set -u
for RAW_FASTQ in $RAW_DATA_DIR/*_R?_001.fastq.gz; do
  echo $RAW_FASTQ
  FASTQ_BASE=$(basename "$RAW_FASTQ" .fastq.gz)
  echo $FASTQ_BASE
  INDEX_FASTQ="${RAW_FASTQ//_R[1-2]_001/_I1_001}"
  $CONDA_BIN/split_libraries_fastq.py \
    -r 999 -n 999 -q 0 -p 0.0001 \
		--sequence_read_fps $RAW_FASTQ \
		--output_dir $TAGGED_FASTQ_DIR/${FASTQ_BASE} \
		--barcode_read_fps $INDEX_FASTQ \
		--mapping_fps $QIIME_MAP_FILE \
		--barcode_type golay_12 \
		--rev_comp_mapping_barcodes \
		--store_demultiplexed_fastq \
		--retain_unassigned_reads
	mv $TAGGED_FASTQ_DIR/${FASTQ_BASE}/seqs.fastq $TAGGED_FASTQ_DIR/${FASTQ_BASE}_tagged.fastq
	mv $TAGGED_FASTQ_DIR/${FASTQ_BASE}/split_library_log.txt $TAGGED_FASTQ_DIR/${FASTQ_BASE}_split_library_log.txt
  rm -r $TAGGED_FASTQ_DIR/${FASTQ_BASE}
done
```
# WORKS TO HERE
```{bash}
ls -ltr $TAGGED_FASTQ_DIR
date
```
# Split fastqs and gzip
```{bash}
set -u
for TAGGED_FASTQ in $TAGGED_FASTQ_DIR/*.fastq; do
  echo $TAGGED_FASTQ
  FASTQ_BASE=$(basename "$TAGGED_FASTQ" _tagged.fastq)
  echo $FASTQ_BASE
  SPLIT_FASTQ_DIR=${SPLIT_FASTQ_BASE}/${FASTQ_BASE}
  echo $SPLIT_FASTQ_DIR
  ${CONDA_BIN}/split_sequence_file_on_sample_ids.py -i $TAGGED_FASTQ \
					 --file_type fastq \
					 --output_dir $SPLIT_FASTQ_DIR

  for FASTQ in $SPLIT_FASTQ_DIR/*.fastq; do
    gzip --stdout $FASTQ > ${FASTQ}.gz
    rm $FASTQ
  done
done
```


# Generate MD5SUMs
```{bash}
set -u
cd ${SPLIT_FASTQ_BASE}
md5sum */*.fastq.gz > $(basename $SPLIT_FASTQ_MD5SUM)
```

# Prep FASTQs for submission
## Copy FASTQs for submission
```{bash}
set -u
mkdir -p $COMBINED_READ_DIR
for READ in R1 R2; do
  echo $READ
  CUR_DIR=${SPLIT_FASTQ_BASE}/Undetermined_S0_L001_${READ}_001
  echo $CUR_DIR
  for FASTQ in $CUR_DIR/*.fastq.gz; do
    # echo $FASTQ
    TARGET_FASTQ=$COMBINED_READ_DIR/$(basename $FASTQ .fastq.gz).${READ}.fastq.gz
    cp $FASTQ $TARGET_FASTQ
  done
done
```

## Generate MD5SUMs for submission
```{bash}
set -u

cd ${COMBINED_READ_DIR}
md5sum *.fastq.gz > $(basename $COMBINED_READ_MD5SUM)
```

## Generate Experiment Spreadsheet for submission
```{r}
md5_both_reads = read_table(combined_read_md5sum, col_names = c("md5sum","filename")) %>%
  mutate(sampleid = str_replace(filename, ".R[12].fastq.gz", "")) %>%
  mutate(read_end = str_extract(filename, "R[12]"))

md5_forward = md5_both_reads %>%
  filter(read_end == "R1") %>%
  select(sample_alias = sampleid, forward_file_name = filename, forward_file_md5 = md5sum )

md5_reverse = md5_both_reads %>%
  filter(read_end == "R2") %>%
  select(sample_alias = sampleid, reverse_file_name = filename, reverse_file_md5 = md5sum )

full_join(md5_forward, md5_reverse, by="sample_alias") %>%
  mutate(instrument_model = "Illumina MiSeq", 
         library_name = "",
         library_source = "METAGENOMIC",
         library_selection = "PCR", 
         library_strategy = "AMPLICON",
         design_description = "",
         library_construction_protocol = "16S rRNA v4 region  was amplified using the EMP 515F and 806R primers",
         insert_size = "253"
         ) %>%
  filter(sample_alias != "Unassigned") %>%
  write_tsv(experiment_table_file)
```



#--------------------------------------------------
TAXONOMY_DIR := $(WORKSPACE_DIR)/taxonomy_refs
SILVA_DB := $(TAXONOMY_DIR)/silva_nr_v123_train_set.fa.gz

$(FULL_PHYLOSEQ_RDS) : $(SPLIT_FASTQS) $(SILVA_DB)
	$(dir_guard)
	# Rscript --no-restore process_fastq_to_counts.R --quality_plots 5 
	# Rscript --no-restore process_fastq_to_counts.R --filter_fastqs
	Rscript --no-restore process_fastq_to_counts.R


$(SILVA_DB) :
	$(dir_guard)
	wget -O $@_tmp "http://benjjneb.github.io/dada2/Training/silva_nr_v123_train_set.fa.gz"
	mv $@_tmp $@


# mkdir -p $SPLIT_FASTQ_BASE

# validate_mapping_file.py --mapping_fp $QIIME_MAP_FILE --output_dir $WORKSPACE_DIR
